File: fpl_app/logic/chip_sequences.py
"""Paired, bounded chip sequences with the same continuation policy.

This is a transparent counterfactual, not another season optimizer.  One
already-solved Wildcard squad and the normal first action receive the same
deterministic continuation search: up to two greedy transfers on deadlines
two to four, screened to 24 position-preserving candidates per transfer.
No extra solver processes or network calls are used.
"""

from __future__ import annotations

from collections import Counter
from dataclasses import dataclass, replace
from functools import lru_cache
from time import monotonic
from typing import Collection, Mapping, Sequence

import pandas as pd

from ..domain.rules import (
    FORMATIONS,
    Chip,
    can_play_chip,
    chip_window,
    free_transfers_next_gameweek,
    selling_price_tenths,
    transfer_points_cost,
)
from .rolling_transfers import DEFAULT_BENCH_WEIGHTS
from .squad_plan import SquadPlanResult
from .strategy_planner import BoundedStrategyResult


MODEL_SCOPE = "paired_bounded_chip_sequences"
CONTINUATION_SHORTLIST = 24
MAX_CONTINUATION_TRANSFERS = 2
ASSUMPTIONS = (
    "same_horizon_weights_and_continuation_policy",
    "fixed_current_prices_with_actual_selling_basis",
    "normal_first_action_from_transfer_roadmap",
    "one_wildcard_squad_not_jointly_optimised_for_bench_boost",
    "at_most_two_greedy_transfers_at_deadlines_two_to_four",
    "continuation_candidates_from_existing_roadmap_and_wildcard_squads",
    "top_24_individual_gain_candidates_per_transfer",
    "one_weighted_point_hurdle_for_each_provisional_transfer",
    "no_transfers_after_deadline_four",
    "one_bench_boost_in_the_current_chip_half_only",
    "future_chip_option_value_not_priced",
    "recalculate_at_every_deadline",
)


class SequenceBudgetExceeded(RuntimeError):
    """The paired comparison must be dropped if either side runs out of time."""


@dataclass(frozen=True)
class _Action:
    event: int
    chip: str | None
    transfer_out_ids: tuple[int, ...]
    transfer_in_ids: tuple[int, ...]
    squad_ids: tuple[int, ...]
    bank_after_tenths: int
    free_transfers_before: int
    free_transfers_next_gameweek: int
    hit_points: int

    def as_dict(self) -> dict[str, object]:
        return {
            "event": self.event,
            "chip": self.chip,
            "transfer_out_ids": list(self.transfer_out_ids),
            "transfer_in_ids": list(self.transfer_in_ids),
            "squad_ids": list(self.squad_ids),
            "bank_after_tenths": self.bank_after_tenths,
            "free_transfers_before": self.free_transfers_before,
            "free_transfers_next_gameweek": self.free_transfers_next_gameweek,
            "hit_points": self.hit_points,
        }


@dataclass(frozen=True)
class _Path:
    sequence_id: str
    label: str
    weighted_net_points: float
    actions: tuple[_Action, ...]

    def as_dict(self, normal_points: float) -> dict[str, object]:
        return {
            "sequence_id": self.sequence_id,
            "label": self.label,
            "weighted_net_points": round(self.weighted_net_points, 3),
            "gain_vs_normal_points": round(self.weighted_net_points - normal_points, 3),
            "total_hit_points": sum(action.hit_points for action in self.actions),
            "actions": [action.as_dict() for action in self.actions],
        }


def _chip_available(chip: Chip, event: int, usage: Mapping[Chip, Sequence[int]]) -> bool:
    used_halves = {
        window.half
        for used_event in usage.get(chip, ())
        for window in [chip_window(chip, used_event)]
        if window is not None
    }
    return can_play_chip(chip, event, used_halves=used_halves)


def evaluate_chip_sequences(
    players: pd.DataFrame,
    roadmap: BoundedStrategyResult,
    wildcard_plan: SquadPlanResult | None,
    *,
    target_event: int,
    usage: Mapping[Chip, Sequence[int]],
    current_squad_ids: Sequence[int],
    bank_tenths: int,
    free_transfers: int,
    eligible_transfer_in_ids: Collection[int],
    deadline: float,
) -> dict[str, object]:
    """Compare matched no-chip, BB, WC and WC-then-BB paths if time permits.

The Wildcard is used only at the current deadline.  Bench Boost candidates
cannot cross the current chip-set expiry.  Scores exclude terminal FT bonuses
and never treat unused chips as worthless: that unpriced option is explicit.
    """

    original = roadmap.best_roadmap
    result: dict[str, object] = {
        "status": "unavailable",
        "horizon": roadmap.horizon,
        "model_scope": MODEL_SCOPE,
        "globally_optimal": False,
        "recalculate_each_deadline": True,
        "original_roadmap_weighted_net_points": round(
            original.weighted_projected_points - original.weighted_hit_cost_points, 3
        ),
        "highest_projected_sequence_id": None,
        "sequences": [],
        "assumptions": list(ASSUMPTIONS),
        "reason": "Den parrede chipanalyse mangler et sikkert tidsbudget.",
    }
    if monotonic() >= deadline:
        return result

    weights = roadmap.gw_weights
    horizon = roadmap.horizon
    rows = {int(row["id"]): row for row in players.to_dict("records")}
    initial_ids = frozenset(int(value) for value in current_squad_ids)
    reference_ids = {
        int(value) for step in roadmap.best_roadmap.steps for value in step.squad_ids
    }
    if wildcard_plan is not None:
        reference_ids.update(int(value) for value in wildcard_plan.squad_ids)
    eligible = frozenset(int(value) for value in eligible_transfer_in_ids) & reference_ids
    positions = {player_id: str(row["pos"]) for player_id, row in rows.items()}
    teams = {player_id: int(row["team_id"]) for player_id, row in rows.items()}
    costs = {player_id: int(row["now_cost"]) for player_id, row in rows.items()}
    ep = {
        offset: {player_id: float(row[f"ep_gw{offset}"]) for player_id, row in rows.items()}
        for offset in range(1, horizon + 1)
    }
    reserve_weights = DEFAULT_BENCH_WEIGHTS
    residual = {
        offset: {
            player_id: ep[offset][player_id]
            * (1.0 - float(row[f"no_show_prob_gw{offset}"]))
            * (reserve_weights[3] if positions[player_id] == "GKP" else sum(reserve_weights[:3]) / 3)
            for player_id, row in rows.items()
        }
        for offset in range(1, horizon + 1)
    }
    formations = [
        {position.value: count for position, count in formation.positions.items()}
        for formation in FORMATIONS
    ]

    def check_budget() -> None:
        if monotonic() >= deadline:
            raise SequenceBudgetExceeded

    @lru_cache(maxsize=12000)
    def lineup(squad: tuple[int, ...], offset: int) -> tuple[float, tuple[int, ...]]:
        """Exact XI/captain score under the roadmap's linear reserve model."""
        check_budget()
        groups = {
            position: sorted(
                (player_id for player_id in squad if positions[player_id] == position),
                key=lambda player_id: (-(ep[offset][player_id] - residual[offset][player_id]), player_id),
            )
            for position in ("GKP", "DEF", "MID", "FWD")
        }
        base = sum(residual[offset][player_id] for player_id in squad)
        best_score = float("-inf")
        best_starters: tuple[int, ...] = ()
        adjusted = {value: ep[offset][value] - residual[offset][value] for value in squad}
        for counts in formations:
            starters = tuple(value for position, count in counts.items() for value in groups[position][:count])
            selected = frozenset(starters)
            base_score = base + sum(adjusted[value] for value in starters)
            for captain in squad:
                score = base_score + ep[offset][captain]
                captain_starters = starters
                if captain not in selected:
                    displaced = groups[positions[captain]][counts[positions[captain]] - 1]
                    score += adjusted[captain] - adjusted[displaced]
                    captain_starters = tuple(value for value in starters if value != displaced) + (captain,)
                ids = tuple(sorted(captain_starters))
                if score > best_score + 1e-9 or (abs(score - best_score) < 1e-9 and ids < best_starters):
                    best_score, best_starters = score, ids
        return best_score, best_starters

    def remaining_score(squad: frozenset[int], offset: int) -> float:
        return sum(
            weights[later - 1] * lineup(tuple(sorted(squad)), later)[0]
            for later in range(offset, horizon + 1)
        )

    def continue_path(
        *,
        sequence_id: str,
        label: str,
        first_squad: frozenset[int],
        first_bank: int,
        first_chip: Chip | None,
        first_out: tuple[int, ...],
        first_in: tuple[int, ...],
    ) -> _Path:
        squad = first_squad
        bank = first_bank
        purchases = {
            player_id: int(rows[player_id]["purchase_price"]) if player_id in initial_ids and player_id not in first_in else costs[player_id]
            for player_id in squad
        }
        ft = free_transfers
        first_hit = transfer_points_cost(len(first_out), ft, active_chip=first_chip)
        next_ft = free_transfers_next_gameweek(ft, len(first_out), active_chip=first_chip)
        actions = [
            _Action(target_event, first_chip.value if first_chip else None, first_out, first_in,
                    tuple(sorted(squad)), bank, ft, next_ft, first_hit)
        ]
        total = weights[0] * (lineup(tuple(sorted(squad)), 1)[0] - first_hit)
        ft = next_ft
        for offset in range(2, horizon + 1):
            check_budget()
            outs: list[int] = []
            ins: list[int] = []
            # The same greedy search is applied to both paths.  Candidate
            # screening is deliberately independent of chip/manager identity.
            if offset <= 4:
                for _ in range(MAX_CONTINUATION_TRANSFERS):
                    current_score = remaining_score(squad, offset)
                    candidate_moves = []
                    for outgoing in sorted(squad - set(ins)):
                        sale = selling_price_tenths(purchases[outgoing], costs[outgoing])
                        for incoming in sorted(eligible - squad - set(outs)):
                            if positions[incoming] != positions[outgoing] or costs[incoming] > bank + sale:
                                continue
                            revised = (squad - {outgoing}) | {incoming}
                            if max(Counter(teams[value] for value in revised).values()) > 3:
                                continue
                            screen_gain = sum(
                                weights[later - 1] * (ep[later][incoming] - ep[later][outgoing])
                                for later in range(offset, horizon + 1)
                            )
                            candidate_moves.append((screen_gain, outgoing, incoming, revised, bank + sale - costs[incoming]))
                    candidate_moves.sort(key=lambda move: (-move[0], move[1], move[2]))
                    best = None
                    best_gain = 0.0
                    incremental_hit = transfer_points_cost(len(outs) + 1, ft) - transfer_points_cost(len(outs), ft)
                    for _, outgoing, incoming, revised, next_bank in candidate_moves[:CONTINUATION_SHORTLIST]:
                        gain = remaining_score(frozenset(revised), offset) - current_score - weights[offset - 1] * incremental_hit
                        # A one-point hurdle protects optionality; this is a
                        # disclosed heuristic, not a calibrated FT valuation.
                        if gain > 1.0 and gain > best_gain + 1e-9:
                            best_gain = gain
                            best = (outgoing, incoming, frozenset(revised), next_bank)
                    if best is None:
                        break
                    outgoing, incoming, squad, bank = best
                    purchases.pop(outgoing)
                    purchases[incoming] = costs[incoming]
                    outs.append(outgoing)
                    ins.append(incoming)
            hit = transfer_points_cost(len(outs), ft)
            next_ft = free_transfers_next_gameweek(ft, len(outs))
            actions.append(_Action(target_event + offset - 1, None, tuple(outs), tuple(ins),
                                   tuple(sorted(squad)), bank, ft, next_ft, hit))
            total += weights[offset - 1] * (lineup(tuple(sorted(squad)), offset)[0] - hit)
            ft = next_ft
        return _Path(sequence_id, label, total, tuple(actions))

    def add_bench_boost(path: _Path) -> _Path | None:
        current_window = chip_window(Chip.BENCH_BOOST, target_event)
        candidates: list[tuple[float, int]] = []
        for offset, action in enumerate(path.actions, start=1):
            window = chip_window(Chip.BENCH_BOOST, action.event)
            if (action.chip is not None or window is None or current_window is None
                    or window.half != current_window.half
                    or not _chip_available(Chip.BENCH_BOOST, action.event, usage)):
                continue
            _, starters = lineup(action.squad_ids, offset)
            bench = set(action.squad_ids) - set(starters)
            gain = sum(ep[offset][value] - residual[offset][value] for value in bench)
            candidates.append((weights[offset - 1] * gain, offset))
        if not candidates:
            return None
        gain, offset = max(candidates, key=lambda candidate: (candidate[0], -candidate[1]))
        # Do not propose using a chip to reduce the modelled score.
        if gain <= 0:
            return None
        actions = list(path.actions)
        actions[offset - 1] = replace(actions[offset - 1], chip=Chip.BENCH_BOOST.value)
        return _Path(f"{path.sequence_id}-bboost", f"{path.label} + Bench Boost GW{target_event + offset - 1}",
                     path.weighted_net_points + gain, tuple(actions))

    try:
        first = original.steps[0]
        normal = continue_path(
            sequence_id="normal", label="Normale transfers",
            first_squad=frozenset(first.squad_ids), first_bank=first.bank_after_tenths,
            first_chip=None, first_out=tuple(move.out_id for move in first.transfers),
            first_in=tuple(move.in_id for move in first.transfers),
        )
        paths = [normal]
        normal_bb = add_bench_boost(normal)
        if normal_bb is not None:
            paths.append(normal_bb)
        if wildcard_plan is not None and _chip_available(Chip.WILDCARD, target_event, usage):
            wc_ids = frozenset(int(value) for value in wildcard_plan.squad_ids)
            sold = initial_ids - wc_ids
            bought = wc_ids - initial_ids
            wc_bank = bank_tenths + sum(int(rows[value]["selling_price"]) for value in sold) - sum(costs[value] for value in bought)
            if wc_bank < 0:
                raise ValueError("Wildcard sequence exceeds actual selling-value budget")
            wildcard = continue_path(
                sequence_id="wildcard", label=f"Wildcard GW{target_event}, derefter transfers",
                first_squad=wc_ids, first_bank=wc_bank, first_chip=Chip.WILDCARD,
                first_out=tuple(sorted(sold, key=lambda value: (positions[value], value))),
                first_in=tuple(sorted(bought, key=lambda value: (positions[value], value))),
            )
            paths.append(wildcard)
            wildcard_bb = add_bench_boost(wildcard)
            if wildcard_bb is not None:
                paths.append(wildcard_bb)
        check_budget()
    except SequenceBudgetExceeded:
        # Never display only the side that happened to finish first.
        return result

    highest = max(paths, key=lambda path: path.weighted_net_points)
    has_wildcard = any(path.sequence_id == "wildcard" for path in paths)
    result.update({
        "status": "ready",
        "highest_projected_sequence_id": highest.sequence_id,
        "sequences": [path.as_dict(normal.weighted_net_points) for path in paths],
        "reason": (
            (
                "Samme afgrænsede transferpolitik er kørt på begge starttrupper. "
                if has_wildcard
                else "Kun normale transfers og eventuel Bench Boost kunne sammenlignes; "
                     "der foreligger ingen tilgængelig, løst Wildcard-trup. "
            )
            +
            "Den højeste prognose er ikke automatisk den bedste chipbeslutning: "
            "værdien af at gemme en chip til senere er ikke prissat. "
            "Wildcardets 15 spillere er ikke særskilt optimeret til Bench Boost."
        ),
    })
    return result

File: lib/chip-sequences.ts
import type { PlannerChipInventoryEntry, PlannerPlayerReference, PlannerStrategyPlan } from "./planner-contract.ts";

export type ChipSequenceId = "normal" | "normal-bboost" | "wildcard" | "wildcard-bboost";
export interface ChipSequenceAction {
  event: number;
  chip: "wildcard" | "bboost" | null;
  transfer_out_ids: number[];
  transfer_in_ids: number[];
  squad_ids: number[];
  bank_after_tenths: number;
  free_transfers_before: number;
  free_transfers_next_gameweek: number;
  hit_points: number;
}
export interface ChipSequence {
  sequence_id: ChipSequenceId;
  label: string;
  weighted_net_points: number;
  gain_vs_normal_points: number;
  total_hit_points: number;
  actions: ChipSequenceAction[];
}
export interface ChipSequenceComparison {
  status: "ready" | "unavailable";
  horizon: number;
  model_scope: "paired_bounded_chip_sequences";
  globally_optimal: false;
  recalculate_each_deadline: true;
  original_roadmap_weighted_net_points: number;
  highest_projected_sequence_id: ChipSequenceId | null;
  sequences: ChipSequence[];
  assumptions: string[];
  reason: string;
}
export interface ChipSequenceContext {
  strategy: PlannerStrategyPlan;
  players: ReadonlyMap<number, PlannerPlayerReference>;
  squad: ReadonlyMap<number, PlannerPlayerReference>;
  selling: ReadonlyMap<number, number>;
  bank: number;
  inventory: PlannerChipInventoryEntry[];
}

const IDS: ChipSequenceId[] = ["normal", "normal-bboost", "wildcard", "wildcard-bboost"];
export const CHIP_SEQUENCE_ASSUMPTIONS = [
  "same_horizon_weights_and_continuation_policy",
  "fixed_current_prices_with_actual_selling_basis",
  "normal_first_action_from_transfer_roadmap",
  "one_wildcard_squad_not_jointly_optimised_for_bench_boost",
  "at_most_two_greedy_transfers_at_deadlines_two_to_four",
  "continuation_candidates_from_existing_roadmap_and_wildcard_squads",
  "top_24_individual_gain_candidates_per_transfer",
  "one_weighted_point_hurdle_for_each_provisional_transfer",
  "no_transfers_after_deadline_four",
  "one_bench_boost_in_the_current_chip_half_only",
  "future_chip_option_value_not_priced",
  "recalculate_at_every_deadline",
];

function fail(message: string): never { throw new TypeError(`chip_sequences: ${message}`); }
function exact(value: unknown, keys: string[]): Record<string, unknown> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) fail("Expected an object");
  const row = value as Record<string, unknown>;
  if (Object.keys(row).length !== keys.length || keys.some((key) => !Object.hasOwn(row, key))) fail("Unexpected fields");
  return row;
}
function numeric(value: unknown, min: number, max: number, integer = false): number {
  if (typeof value !== "number" || !Number.isFinite(value) || value < min || value > max || (integer && !Number.isSafeInteger(value))) fail("Invalid number");
  return value;
}
function text(value: unknown, maximum: number): string {
  if (typeof value !== "string" || !value.trim() || value.length > maximum || /[\u0000-\u001f\u007f]/.test(value)) fail("Invalid text");
  return value;
}
function near(actual: number, expected: number): void {
  if (Math.abs(actual - expected) > 0.003) fail("Scores do not reconcile");
}
function ids(value: unknown, maximum: number): number[] {
  if (!Array.isArray(value) || value.length > maximum) fail("Invalid player list");
  const result = value.map((id) => numeric(id, 1, 100_000, true));
  if (new Set(result).size !== result.length) fail("Duplicate player IDs");
  return result;
}
function sameSet(actual: number[], expected: Iterable<number>): void {
  const other = [...expected].sort((a, b) => a - b);
  const sorted = [...actual].sort((a, b) => a - b);
  if (sorted.length !== other.length || sorted.some((value, index) => value !== other[index])) fail("Player transitions do not reconcile");
}
function validSquad(squad: number[], players: ChipSequenceContext["players"]): void {
  if (squad.length !== 15) fail("Squad must contain fifteen players");
  const positions = new Map<string, number>();
  const teams = new Map<string, number>();
  for (const id of squad) {
    const player = players.get(id);
    if (!player) fail("Unknown squad player");
    positions.set(player.position, (positions.get(player.position) ?? 0) + 1);
    teams.set(player.team, (teams.get(player.team) ?? 0) + 1);
  }
  for (const [position, count] of Object.entries({ GKP: 2, DEF: 5, MID: 5, FWD: 3 })) {
    if (positions.get(position) !== count) fail("Illegal squad position quota");
  }
  if ([...teams.values()].some((count) => count > 3)) fail("More than three players from one club");
}

export function parseChipSequences(value: unknown, context: ChipSequenceContext): ChipSequenceComparison {
  const row = exact(value, ["status", "horizon", "model_scope", "globally_optimal", "recalculate_each_deadline", "original_roadmap_weighted_net_points", "highest_projected_sequence_id", "sequences", "assumptions", "reason"]);
  if (row.status !== "ready" && row.status !== "unavailable") fail("Unsupported status");
  if (row.model_scope !== "paired_bounded_chip_sequences" || row.globally_optimal !== false || row.recalculate_each_deadline !== true) fail("Unsupported model scope");
  const horizon = numeric(row.horizon, 6, 10, true);
  if (horizon !== context.strategy.horizon || context.strategy.gameweek_window.length !== horizon || context.strategy.gw_weights.length !== horizon) fail("Horizon does not match the roadmap");
  const originalPoints = numeric(row.original_roadmap_weighted_net_points, -1_000, 10_000);
  near(originalPoints, context.strategy.weighted_projected_points - context.strategy.weighted_hit_cost_points);
  if (!Array.isArray(row.assumptions) || row.assumptions.length !== CHIP_SEQUENCE_ASSUMPTIONS.length
    || row.assumptions.some((assumption, index) => assumption !== CHIP_SEQUENCE_ASSUMPTIONS[index])) fail("Unsupported assumptions");
  if (!Array.isArray(row.sequences) || row.sequences.length > 4) fail("Invalid sequence list");
  const first = context.strategy.steps[0];
  const target = context.strategy.gameweek_window[0];
  const currentHalf = target <= 19 ? 1 : 2;
  const available = (chip: "wildcard" | "bboost", event: number) => {
    const inventory = context.inventory.find((item) => item.chip === chip);
    return !!inventory?.available_for_target && (event <= 19 ? 1 : 2) === currentHalf
      && !inventory.used_events.some((used) => (used <= 19 ? 1 : 2) === currentHalf);
  };
  const sequences = row.sequences.map((value): ChipSequence => {
    const sequence = exact(value, ["sequence_id", "label", "weighted_net_points", "gain_vs_normal_points", "total_hit_points", "actions"]);
    if (!IDS.includes(sequence.sequence_id as ChipSequenceId)) fail("Unknown sequence");
    const id = sequence.sequence_id as ChipSequenceId;
    if (!Array.isArray(sequence.actions) || sequence.actions.length !== horizon) fail("Incomplete sequence");
    let previous = new Set(context.squad.keys());
    validSquad([...previous], context.players);
    let bank = numeric(context.bank, 0, 10_000, true);
    let ft = first.free_transfers_before;
    const selling = new Map(context.selling);
    let wildcards = 0;
    let boosts = 0;
    const actions = sequence.actions.map((value, index): ChipSequenceAction => {
      const action = exact(value, ["event", "chip", "transfer_out_ids", "transfer_in_ids", "squad_ids", "bank_after_tenths", "free_transfers_before", "free_transfers_next_gameweek", "hit_points"]);
      const event = numeric(action.event, 1, 38, true);
      if (event !== context.strategy.gameweek_window[index]) fail("Wrong sequence event");
      if (action.chip !== null && action.chip !== "wildcard" && action.chip !== "bboost") fail("Unsupported chip");
      const chip = action.chip;
      if (chip !== null && !available(chip, event)) fail("Unavailable chip or expired chip half");
      if (chip === "wildcard") { wildcards++; if (index !== 0 || !id.startsWith("wildcard")) fail("Wildcard must be the first action"); }
      if (chip === "bboost") { boosts++; if (!id.endsWith("-bboost")) fail("Unexpected Bench Boost"); }
      const outgoing = ids(action.transfer_out_ids, index === 0 ? 15 : index < 4 ? 2 : 0);
      const incoming = ids(action.transfer_in_ids, outgoing.length);
      if (outgoing.length !== incoming.length || outgoing.some((value) => incoming.includes(value))) fail("Invalid transfers");
      const revised = new Set(previous);
      for (let move = 0; move < outgoing.length; move++) {
        const out = outgoing[move], next = incoming[move];
        const outgoingPlayer = context.players.get(out), incomingPlayer = context.players.get(next);
        if (!previous.has(out) || previous.has(next) || !outgoingPlayer || !incomingPlayer
          || outgoingPlayer.position !== incomingPlayer.position) fail("Transfer does not match the owned squad and position");
        const sale = selling.get(out);
        if (sale === undefined) fail("Missing selling price");
        bank += numeric(sale, 0, 500, true) - incomingPlayer.price_tenths;
        revised.delete(out); revised.add(next);
        selling.delete(out); selling.set(next, incomingPlayer.price_tenths);
      }
      const squad = ids(action.squad_ids, 15);
      sameSet(squad, revised); validSquad(squad, context.players);
      const reportedBank = numeric(action.bank_after_tenths, 0, 10_000, true);
      if (reportedBank !== bank) fail("Transfer bank does not reconcile with actual selling values");
      const ftBefore = numeric(action.free_transfers_before, 0, 5, true);
      const ftAfter = numeric(action.free_transfers_next_gameweek, 0, 5, true);
      const hit = numeric(action.hit_points, 0, 80, true);
      if (ftBefore !== ft || ftAfter !== (chip === "wildcard" ? ft : Math.min(5, Math.max(0, ft - outgoing.length) + 1))
        || hit !== (chip === "wildcard" ? 0 : 4 * Math.max(0, outgoing.length - ft))) fail("Invalid free transfers or hit cost");
      if (index === 0 && id.startsWith("normal")) {
        sameSet(squad, first.squad_ids);
        sameSet(outgoing, first.transfers.map((move) => move.out_id));
        sameSet(incoming, first.transfers.map((move) => move.in_id));
        if (reportedBank !== first.bank_after_tenths || ftAfter !== first.free_transfers_next_gameweek || hit !== first.hit_points) fail("Normal first action differs from the roadmap");
      }
      previous = revised; ft = ftAfter;
      return { event, chip, transfer_out_ids: outgoing, transfer_in_ids: incoming, squad_ids: squad,
        bank_after_tenths: reportedBank, free_transfers_before: ftBefore, free_transfers_next_gameweek: ftAfter, hit_points: hit };
    });
    if (wildcards !== (id.startsWith("wildcard") ? 1 : 0) || boosts !== (id.endsWith("-bboost") ? 1 : 0)) fail("Chip sequence does not match its identity");
    const totalHits = numeric(sequence.total_hit_points, 0, 200, true);
    if (totalHits !== actions.reduce((sum, action) => sum + action.hit_points, 0)) fail("Total hit cost does not reconcile");
    return { sequence_id: id, label: text(sequence.label, 160), weighted_net_points: numeric(sequence.weighted_net_points, -1_000, 10_000),
      gain_vs_normal_points: numeric(sequence.gain_vs_normal_points, -10_000, 10_000), total_hit_points: totalHits, actions };
  });
  if (new Set(sequences.map((sequence) => sequence.sequence_id)).size !== sequences.length) fail("Duplicate sequences");
  let highest: ChipSequenceId | null = null;
  if (row.status === "unavailable") {
    if (sequences.length !== 0 || row.highest_projected_sequence_id !== null) fail("Unavailable comparison contains results");
  } else {
    const normal = sequences.find((sequence) => sequence.sequence_id === "normal");
    if (!normal || !IDS.includes(row.highest_projected_sequence_id as ChipSequenceId)) fail("Missing normal baseline or highest sequence");
    highest = row.highest_projected_sequence_id as ChipSequenceId;
    const winner = sequences.find((sequence) => sequence.sequence_id === highest);
    if (!winner || sequences.some((sequence) => sequence.weighted_net_points > winner.weighted_net_points + 0.003)) fail("Incorrect highest projected sequence");
    for (const sequence of sequences) {
      near(sequence.gain_vs_normal_points, sequence.weighted_net_points - normal.weighted_net_points);
      if (!sequence.sequence_id.endsWith("-bboost")) continue;
      const base = sequences.find((candidate) => candidate.sequence_id === sequence.sequence_id.replace("-bboost", ""));
      if (!base || sequence.weighted_net_points < base.weighted_net_points) fail("Bench Boost reduced the modelled score");
      for (let i = 0; i < horizon; i++) {
        const a = sequence.actions[i], b = base.actions[i];
        if (a.chip === "bboost" && b.chip !== null) fail("Two chips on one deadline");
        if (JSON.stringify({ ...a, chip: b.chip }) !== JSON.stringify(b)) fail("Bench Boost changed the underlying transfer path");
      }
    }
  }
  return { status: row.status, horizon, model_scope: "paired_bounded_chip_sequences", globally_optimal: false,
    recalculate_each_deadline: true, original_roadmap_weighted_net_points: originalPoints, highest_projected_sequence_id: highest,
    sequences, assumptions: [...CHIP_SEQUENCE_ASSUMPTIONS], reason: text(row.reason, 800) };
}
